gateway/governance.go

```diff
@@ -85,21 +85,32 @@ func (g *MagalixGateway) SetAuditCommandHandler(handler agent.AuditCommandHandle
 }
 
 func (g *MagalixGateway) SendAuditResults(auditResults []*agent.AuditResult) error {
-	noOfBatches := int(math.Ceil(float64(len(auditResults)) / float64(auditResultsBatchMaxSize)))
-	lastBatchSize := len(auditResults) % auditResultsBatchMaxSize
-	for i := 0; i < noOfBatches; i++ {
-		start := i * auditResultsBatchMaxSize
-		var end int
-		if i == noOfBatches-1 && lastBatchSize > 0 {
-			end = start + lastBatchSize
-		} else {
-			end = start + auditResultsBatchMaxSize
-		}
-		g.SendAuditResultsBatch(auditResults[start:end])
+	for _, auditResult := range auditResults {
+		g.auditResultChan <- auditResult
 	}
 	return nil
 }
 
+func (g *MagalixGateway) SendAuditResultsWorker(ctx context.Context) error {
+	timer := time.NewTicker(auditResultsBatchExpiry)
+	for {
+		select {
+		case result := <-g.auditResultChan:
+			g.auditResultsBuffer = append(g.auditResultsBuffer, result)
+			if len(g.auditResultsBuffer) == cap(g.auditResultsBuffer) {
+				g.SendAuditResultsBatch(g.auditResultsBuffer)
+				g.auditResultsBuffer = g.auditResultsBuffer[:0]
+				timer.Reset(auditResultsBatchExpiry)
+			}
+		case <-timer.C:
+			if len(g.auditResultsBuffer) > 0 {
+				g.SendAuditResultsBatch(g.auditResultsBuffer)
+				g.auditResultsBuffer = g.auditResultsBuffer[:0]
+			}
+		}
```

Comment:
Won't you need to reset the timer here too?
How about having a flush method that does these procedures? send the batch and clear the buff and reset the timer?

Reply:
This is a ticker we shouldn't need to reset it, it will run per the interval specified

Reply:
should we be handling the context whenever it gets closed? so that we send the data before closing the agent maybe?

Reply:
Yes that would be better indeed